driver/vhci_ude/vhci_urbr_store.c

```diff
@@ -50,6 +50,10 @@ store_urbr_partial(WDFREQUEST req_read, purb_req_t urbr)
 
 	TRD(READ, "Enter: urbr: %!URBR!", urbr);
 
+	/* This request is sometimes NULL, which causes driver/system failure, if used. */
+	if (urbr->req == NULL)
```

Comment:
I have checked that urbr with no req(select conf/intf, unlink) is not invoked. Can you clarify a reproducible error case?

Reply:
@cezzane - i surely had that case, however it might be the side effect of some other error.

Reply:
@cezanne : i will try to repoproduce this case, but it may take some time...

Reply:
@cezanne : Here is how the system crashes occasionally, when i comment out this change. A appended the '!analyze -v' command output of the 'minidump' and a picture of the stack frame with 'urbr->req' value.
[Debugger-minidump-analyze.txt](https://github.com/cezanne/usbip-win/files/5975687/Debugger-minidump-analyze.txt)
![Debugger-minidump-screen-view](https://user-images.githubusercontent.com/7649004/107850615-fe93fa00-6e03-11eb-8d05-feb5eb30e2df.png)

Reply:
@spog: Thanks for your efforts.
It would have been better that I could browse whole fields of urbr. :cry: 
 
Maybe that urbr had been already broken by `complete_urbr()`. 
After some inspection, I found that setting `urbr_sent_partial` as NULL is required as follows:

```
diff --git a/driver/vhci_ude/vhci_read.c b/driver/vhci_ude/vhci_read.c
index 82f363b..60fa747 100644
--- a/driver/vhci_ude/vhci_read.c
+++ b/driver/vhci_ude/vhci_read.c
@@ -95,6 +95,7 @@ read_vusb(pctx_vusb_t vusb, WDFREQUEST req)
                BOOLEAN unmarked;
                RemoveEntryListInit(&urbr->list_all);
                unmarked = unmark_cancelable_urbr(urbr);
+               vusb->urbr_sent_partial = NULL;
                WdfSpinLockRelease(vusb->spin_lock);
 
                if (unmarked)
```

Could you retry?

Reply:
I've added your change and restart my test. When i already thought, that crash won't happen again, the system crashed. Here is a picture and the minidump file (i hope it is more useful than the picture:)) 
![030121-36343-01](https://user-images.githubusercontent.com/7649004/109482430-478fb380-7a7e-11eb-9f95-14bf5ad8f213.png)

[030121-36343-01.zip](https://github.com/cezanne/usbip-win/files/6060740/030121-36343-01.zip)

Reply:
@spog: Thanks for your kind capture. 
There was an urbr which had an invalidated req. After inspection, I found the case when BSOD cloud happen. 
WDF request might be cancelled while a partial urbr of the request is stored due to coarse locking.

Could you check 2c593a543eae09c750de2452ead15c4037c9ab16 ?

Reply:
I am testing your latest 'dev' branch and looks promising, with occasional dead-locks of my usb-audio device, which was the case already before. Previously i managed to resolve this issue by interrupting (ctrl-c) the running 'usbip' process, re-attached the device and resume testing. Now, if i try to detach the 'attached' port 0 i allways fail. Then i tried to detach device immediatly after attaching (no deadlock) and got another BSOD (see attachment).

I will continue to play the audio device occasionally to see if anything unexpected happens.
[030221-45312-01.zip](https://github.com/cezanne/usbip-win/files/6071471/030221-45312-01.zip)

Reply:
@spog: I can't get exact information about BSOD from your minidump but understand there's still BSOD while detaching.
Anyway, I'll resolve this conversation because `if (urb->req == NULL)` can be removed by 2c593a543eae09c750de2452ead15c4037c9ab16.

Reply:
ok. Regarding new BSOD, i suspect that the spinlock tried to be locked befor it has been created/initialized.

Reply:
@cezanne: just commenting here to use the context above regarding new BSOD at immediate 'detach'. Namely i found out that the following change prevent BSOD to occure:

```
--- a/driver/vhci_ude/vhci_read.c
+++ b/driver/vhci_ude/vhci_read.c
@@ -143,7 +143,7 @@ io_read(_In_ WDFQUEUE queue, _In_ WDFREQUEST req, _In_ size_t len)
        svusb = TO_SAFE_VUSB(WdfRequestGetFileObject(req));
        vusb = svusb->vusb;
 
-       if (vusb->invalid) {
+       if ((vusb->vhci == NULL) || (vusb->invalid)) {
                TRD(READ, "vusb disconnected: port: %u", vusb->port);
                status = STATUS_DEVICE_NOT_CONNECTED;
        }
```

This probably isn't the correct solution and the question remains, why 'vusb->invalid' isn't TRUE after 'detach'? Hope this helps.

Reply:
@spog: :+1: Thanks for your finding.
Maybe, vusb seems to have been already destroyed. 
This means that detaching by `plugout_vusb()` cannot prevent `io_read()`. I'll inspect more..

Reply:
@cezanne: You are welcome and thank you.